firewood/src/merkle/stream.rs

```diff
@@ -7,399 +7,471 @@ use crate::{
-                        let value = value.to_vec();
+                        IterationNode::Visited {
+                            ref key,
+                            ref mut children_iter,
+                        } => {
+                            // We returned `node` already. Visit its next child.
+                            let Some((child_addr, pos)) = children_iter.next() else {
+                                // We visited all this node's descendants. Go back to its parent.
+                                continue;
+                            };
+
+                            // Visit the child next.
+                            let child = merkle
+                                .get_node(child_addr)
+                                .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                            let partial_path = match child.inner() {
+                                NodeType::Branch(branch) => branch.path.iter().copied(),
+                                NodeType::Leaf(leaf) => leaf.path.iter().copied(),
+                                NodeType::Extension(_) => {
+                                    panic!("extension nodes shouldn't exist")
+                                }
+                            };
+
+                            let child_key: Box<[u8]> = key
+                                .iter()
+                                .copied()
+                                .chain(once(pos))
+                                .chain(partial_path)
+                                .collect();
+
+                            // There may be more children of this node to visit.
+                            iter_stack.push(iter_node);
+
+                            let child = merkle
+                                .get_node(child_addr)
+                                .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                            iter_stack.push(IterationNode::Unvisited {
+                                key: child_key,
+                                node: child,
+                            });
+                            return self.poll_next(_cx);
+                        }
+                    }
+                }
+                Poll::Ready(None)
+            }
+        }
+    }
+}
 
-                        Ok((std::mem::take(key), value))
-                    });
+// TODO comment
+fn get_iterator_intial_state<'a, S: ShaleStore<Node> + Send + Sync, T>(
+    merkle: &'a Merkle<S, T>,
+    root_node: DiskAddress,
+    key: &[u8],
+) -> Result<MerkleNodeStreamState<'a>, api::Error> {
+    // Invariant: `node`'s key is a prefix of `key`.
+    let mut node = merkle
+        .get_node(root_node)
+        .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-                    visited_node_path.push((found_node, 0));
+    // Invariant: This is `node`'s address at the start
+    // of each loop iteration.
+    let mut node_addr = root_node;
```

Comment:
Can you use `node.as_ptr()` instead? Or is there a time when `node_addr` does not point to `node`?

Reply:
removed `node_addr` -- not needed